internal/controller/helmchart_controller.go

```diff
@@ -536,7 +541,8 @@ func (r *HelmChartReconciler) buildFromHelmRepository(ctx context.Context, obj *
 		// this is needed because otherwise the credentials are stored in ~/.docker/config.json.
 		// TODO@souleb: remove this once the registry move to Oras v2
 		// or rework to enable reusing credentials to avoid the unneccessary handshake operations
-		registryClient, credentialsFile, err := r.RegistryClientGenerator(clientOpts.RegLoginOpt != nil)
+		shouldLogin := len(clientOpts.RegLoginOpts) > 0 && clientOpts.RegLoginOpts[0] != nil
```

Comment:
it's `nil` for public repositories actually. Because when we initialise the clientOpts the slice is `nil` and when there are no authentication, we never allocate.

Reply:
Can't this be prevented by doing something like the following in `client_opts.go`?

```go
		loginOpt, err := registry.NewLoginOption(hrOpts.Authenticator, hrOpts.Keychain, url)
		if err != nil {
			return nil, "", err
		}
		if loginOpt != nil {
			hrOpts.RegLoginOpts = append(hrOpts.RegLoginOpts, loginOpt)
		}
```

Reply:
In addition, it may actually be better to move this to a method on the `ClientOpts` instead of making use of a boolean variable.

Reply:
You could still do the following?

```go
func (o ClientOpts) MustLoginToRegistry() bool {
    return len(clientOpts.RegLoginOpts) > 0
}
```

in combination with:

```go
r.RegistryClientGenerator(clientOpts.TlsConfig, clientOpts.MustLoginToRegistry())
```

(of which I would think the latter should actually be a private method?)